## Download the transformers package - Bert pretrained model access (trained on SQuAD, open domain) 
#### As per arxiv:1810, 24 May 2019, paper - BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding

In [1]:
!pip install transformers  

     |████████████████████████████████| 573kB 3.5MB/s 
     |████████████████████████████████| 890kB 50.4MB/s 
     |████████████████████████████████| 3.7MB 41.1MB/s 
     |████████████████████████████████| 1.0MB 41.7MB/s 
  Created wheel for sacremoses: filename=sacremoses-0.0.41-cp36-none-any.whl size=893334 sha256=9108ccc67b1d04960ce5464f9557550187bac1db1d22bcc1efd3855e101db252
  Stored in directory: /root/.cache/pip/wheels/22/5a/d4/b020a81249de7dc63758a34222feaa668dbe8ebfe9170cc9b1
Successfully built sacremoses


In [0]:
import torch
from transformers import *

In [3]:
pretrained_weights = 'bert-base-uncased'
berttokenizer= BertTokenizer.from_pretrained(pretrained_weights)
model = BertModel.from_pretrained(pretrained_weights, 
                                  output_hidden_states=True, 
                                  output_attentions= True)

In [4]:
print(berttokenizer.encode("BERT encoding is used in place of textual encoding in BiDAF"))

[101, 14324, 17181, 2003, 2109, 1999, 2173, 1997, 25304, 17181, 1999, 7226, 10354, 102]


In [5]:
print(berttokenizer.encode("BERT encoding is used in place of textual encoding in BiDAF", add_special_tokens=False))

[14324, 17181, 2003, 2109, 1999, 2173, 1997, 25304, 17181, 1999, 7226, 10354]


#### In the example above that included 101, 102, the two special tokens [CLS] and [SEP] (i guess!) - total tokens= 14
#### While print statement add_special_tokens=False, will not have special tokens, total tokens = 12

In [0]:
input_ids = torch.tensor([berttokenizer.encode("BERT encoding is used in place of textual encoding in BiDAF",
                                               add_special_tokens=False)])
all_hidden_states, all_attentions = model(input_ids)[-2:]

In [7]:
all_hidden_states[0].shape

torch.Size([1, 12, 768])

### The all_hidden_states tuple as above, contains the contextual embeddings for all 12 tokens (no special tokens added) in this example.  